add a doc for QueueingHint (#43813)
kubernetes/website: content/en/docs/concepts/scheduling-eviction/scheduling-framework.md

be unschedulable or if there is an internal error. The Pod will be returned to
the queue and retried.

## Interfaces

The following picture shows the scheduling context of a Pod and the interfaces

> that the scheduling framework exposes. In this picture "Filter" is
> equivalent to "Predicate" and "Scoring" is equivalent to "Priority function".

Reviewer: ```suggestion
that the scheduling framework exposes.
```

It's been a while since we last used the terms "predicate" and "priority". We can get rid of this sentence to avoid confusion.

One plugin may implement multiple interfaces to perform more complex or
stateful tasks.

> Some interfaces are called as extension points which can be controled through 

Reviewer: ```suggestion
Some interfaces match the scheduler extension points which can be configured through 
```

For more details about how internal scheduler queues work, read
[Scheduling queue in kube-scheduler](https://github.com/kubernetes/community/blob/f03b6d5692bd979f07dd472e7b6836b2dad0fd9b/contributors/devel/sig-scheduling/scheduler_queues.md).

### EnqueueExtension

EnqueueExtension is the interface where the plugin can control 
whether to retry scheduling of Pods rejected by this plugin, based on changes in the cluster.
Plugins that implement PreEnqueue, PreFilter, Filter, Reserve or Permit should implement this interface.

#### QueueingHint

{{< feature-state for_k8s_version="v1.28" state="beta" >}}

QueueingHint is a callback function for deciding whether a Pod can be requeued to the active queue or backoff queue. 

> It's executed every time a certain kind of events happen,
> and when the QueueingHint finds that the event might make the Pod schedulable, 

Reviewer: ```suggestion
It's executed every time a certain kind of event or change happens in the cluster.
When the QueueingHint finds that the event might make the Pod schedulable, 
```

the Pod is put into the active queue or the backoff queue
so that the scheduler will retry the scheduling of the Pod.

{{< note >}}

> QueueingHint is a beta-level field and is enabled by default in 1.28. 

Reviewer: ```suggestion
QueueingHint is a beta-level feature and is enabled by default in 1.28. 
```
(it's not an API field)

Reviewer 2: This isn't quite right either. Try something like “Queuing hint evaluation during scheduling is a beta-level feature&hellip;”, or “the QueueingHint scheduler frobnication is a beta feature&hellip;” (change frobnication to a suitable term).

Author: I go with:
QueueingHint evaluation during scheduling is a beta-level feature ..

Pod is successfully bound. This is the end of a binding cycle, and can be used
to clean up associated resources.

### EnqueueExtension

> EnqueueExtension is the interface where the plugin can have control how the scheduling of Pods rejected by this plugin are retried.

Reviewer 3: How about 

```suggestion
EnqueueExtension is the interface where the plugin can control the retry behavior
when the Pods is rejected.
```

Author: Well, this suggestion would lose one important thing -- it controls the retry of the scheduling of Pods **rejected by this plugin are retried.**
Any other good idea to make it better?

Reviewer: ```suggestion
EnqueueExtension is the interface where the plugin can control whether to retry scheduling of Pods rejected by this plugin, based on changes in the cluster.
```

PreEnqueue, PreFilter, Filter, Reserve and Permit plugins are supposed to implement this interface.

{{< feature-state for_k8s_version="v1.28" state="beta" >}}

QueueingHint is a part of EnqueueExtension. 

> It's callback functions to decide if a Pod can be requeued to activeQ/backoffQ. 

Reviewer 3: ```suggestion
It contains callback functions for deciding whether a Pod can be requeued to the active queue or backoff queue. 
```

Author: Looks good except QueueingHint itself is a callback function -- if you say QueueingHint contains callback function, it's kind of weird.

> It's executed every time certain kind of events happens, 

Reviewer 3: ```suggestion
It's executed every time a certain kind of events happen,
```

and when the QueueingHint finds that the event might make the Pod schedulable, 

> the scheduling queue requeues the Pod into activeQ/backoffQ so that the scheduler will retry the scheduling of the Pod.

Reviewer 3: ```suggestion
the Pod is put into the active queue or the backoff queue
so that the scheduler will retry the scheduling of the Pod.
```

{{< note >}}

> QueueingHint is a beta-level field and enabled by default in 1.28. 

Reviewer 3: ```suggestion
QueueingHint is a beta-level field and is enabled by default in 1.28. 
```

Pod is successfully bound. This is the end of a binding cycle, and can be used
to clean up associated resources.

> ### EnqueueExtension

Reviewer: This should go after PreEnqueue.

I wonder if we should rename `## Extension points` to `## Interfaces`, so that it's not confused with https://kubernetes.io/docs/reference/scheduling/config/#extension-points, which actually form part of the API.

Author: Sounds good, I rename this along with some wording changes in the doc.

EnqueueExtension is the interface where the plugin can have control how the scheduling of Pods rejected by this plugin are retried.

> PreEnqueue, PreFilter, Filter, Reserve and Permit plugins are supposed to implement this interface.

Reviewer: ```suggestion
Plugins that implement PreEnqueue, PreFilter, Filter, Reserve or Permit should implement this interface.
```

> {{< feature-state for_k8s_version="v1.28" state="beta" >}}

Reviewer: ```suggestion
#### QueueingHint

{{< feature-state for_k8s_version="v1.28" state="beta" >}}
```

> QueueingHint is a part of EnqueueExtension. 
> It's a callback function for deciding whether a Pod can be requeued to the active queue or backoff queue. 

Reviewer: ```suggestion
QueueingHint is a callback function for deciding whether a Pod can be requeued to the active queue or backoff queue. 
```
